src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -309,6 +311,38 @@ object Task {
     override lazy val reservationId: String = instanceId.idString
   }
 
+  /**
+    * Identifier for app or pod task with and attempt count and type prefix.
+    *
+    * The ids match [[Task.TaskIdWithIncarnation.regex]] and include a launch attempt.
+    *
+    * Examples:
+    *  - "$tiwi$.myGroup_myApp.marathon-b6ff5fa5-7714-11e7-a55c-5ecf1c4671f6.$anon.1"
+    *  - "$tiwi$.myGroup_myApp.instance-b6ff5fa5-7714-11e7-a55c-5ecf1c4671f6.$anon.3"
+    *  - "$tiwi$.myGroup_myApp.marathon-b6ff5fa5-7714-11e7-a55c-5ecf1c4671f6.rails.2"
+    *  - "$tiwi$.myGroup_myApp.instance-b6ff5fa5-7714-11e7-a55c-5ecf1c4671f6.rails.42"
+    *
+    * @param instanceId Identifies the instance the task belongs to.
+    * @param containerName If set identifies the container in the pod. Defaults to [[Task.Id.Names.anonymousContainer]].
+    * @param attempt Counts how often a task has been launched on a specific reservation.
+    */
+  case class TaskIdWithIncarnation(val instanceId: Instance.Id, val containerName: Option[String], attempt: Long) extends Id {
+
+    // A stringifed version of the id.
+    override val idString = TaskIdWithIncarnation.typePrefix + "." + instanceId.idString + "." + containerName.getOrElse(Id.Names.anonymousContainer) + "." + attempt
+
+    // Quick access to the underlying run spec identifier of the task.
+    override lazy val runSpecId: PathId = instanceId.runSpecId
+
+    override lazy val reservationId: String = instanceId.idString
+
+    def increaseAttempt(): TaskIdWithIncarnation = this.copy(attempt = attempt + 1L)
+  }
+  object TaskIdWithIncarnation {
+    val typePrefix = "$tiwi$"
+    val regex = """^\$tiwi\$\.(.+)\.(instance-|marathon-)([^_\.]+)[\._]([^_\.]+)\.(\d+)$""".r
```

Comment:
What if container name will contain dots? is such case filtered somewhere else?

Reply:
I don't think so. So you mean the container `$tiwi$.`?

Reply:
To be honest I think these `$tiwi$...` taksIds look kind of ugly. I like current current `group_app_UUID.N` pattern.

Reply:
At the end it would be `group_app_-marathon-UUIID.$anon.$tiwi$.N`

Reply:
Why not just use a version? Start somewhere. `$v2` is fine, or any other arbitrary number as long as we increment the next one. An integer based taskId version is IMO preferable to a named version.

Apart from that, I guess a purist would ask for a UUID and sstore incarnation count and anything else in metdata, since serializing metadata in the ID is kind of weird.